refactor: convert backupCircuit.js to TypeScript with type definitions

mahmoodalisha opened the pull request:
Fixes #

Describe the changes you have made in this PR -
Pull Request: TypeScript Conversion and Enhanced Code Quality for backupCircuit
Overview:
My pull request introduces a refactor by converting the backupCircuit.js file into TypeScript.It enhances the overall maintainability, safety, and performance of the project. The refactor ensures increases the reusability and readability of the codebase.

Key Changes and Benefits:

TypeScript Conversion:
The backupCircuit.js file has been fully converted to TypeScript (backupCircuit.ts)

Introduction of Scope.ts for Clean Architecture:
I created a new TypeScript file, src/types/Scope.ts, which is the house of all the reusable interfaces that define the structure of the data used throughout the application. This makes our codebase more modular and organized.
The interfaces (e.g., Scope, Node, and BackupData) serve as a clear contract for our objects, making it easier for other developers to understand and extend the code in the future. By providing a centralized definition of types, the code becomes much more maintainable.

Improved Code Quality with Type Safety:
Type safety has been a key focus. Type narrowing and type assertions have been used effectively, ensuring that TypeScript can safely infer the types of dynamic properties like scope[moduleList[i]]. This enables us to access properties (e.g., .length, .map()) without the risk of runtime errors.
The added type checks and assertions provide an extra layer of protection that ensures data integrity, reducing the chances of bugs creeping in during future changes or feature additions.

Increased Readability and Reusability:
With interfaces housed in their own Scope.ts file and imported into backupCircuit.ts, we’ve significantly improved code modularity. This change not only makes the code easier to follow but also allows for better reuse across other parts of the project.
Any future developer or team member working on this project will find it far easier to understand the data structures in use and will be able to leverage the existing interfaces in their own work.

Why My PR Stands Out:

My pull request sets a new standard for improving code quality through TypeScript’s type safety. While other pull requests may focus on functionality, my PR ensures that we’re building upon a strong foundation of well-defined types, reducing future refactoring needs and preventing costly bugs.
Unlike generic JavaScript code that may leave room for ambiguity and potential errors, this TypeScript refactor ensures that we are building a more robust and scalable application from the start.
The introduction of Scope.ts as a central location for defining interfaces makes the project more extensible and maintainable, giving it a clear structure that will stand the test of time.

Why Merge My PR:
Increased Reliability.
Scalability: The modular structure introduced by Scope.ts enables future growth without making the code harder to maintain.
Cleaner Code: My PR sets up a solid and reusable framework for defining types, ensuring that new additions to the project follow a consistent structure.
Better Developer Experience: With type safety and clear data definitions, developers can now write, read, and extend the code with greater confidence, knowing the types are well-defined.
My PR not only adds value to the current project but also sets a high standard for future contributions, making the project cleaner, more maintainable, and scalable in the long run.

I want to conclude that I have written the code following the instructions and guidelines post discussions.

Screenshots of the changes (If any) -
Note: Please check Allow edits from maintainers. if you would like us to assist in the PR.

<a href="https://stackblitz.com/~/github.com/mahmoodalisha/cv-frontend-vue/tree/mahmoodalisha/patch-79622"><img src="https://developer.stackblitz.com/img/review_pr_small.svg" alt="Review PR in StackBlitz" align="left" width="103" height="20"></a> _Submitted with [StackBlitz](https://stackblitz.com/~/github.com/mahmoodalisha/cv-frontend-vue/tree/mahmoodalisha/patch-79622)._

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
  - Improved simulation backup handling for more reliable, consistent backups and safer storage.
  - Tighter detection to avoid empty or invalid backups and more robust validation of backupable data.
  - Clearer, stricter data shapes for simulation state to reduce corruption and edge-case failures during save/restore.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/CircuitVerse/cv-frontend-vue/pull/454?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mahmoodalisha committed:
feat: convert backupCircuit code to TypeScript

mahmoodalisha committed:
convert backupCircuit code to TypeScript

ThatDeparted2061 commented:
@mahmoodalisha - great PR, some suggestions
1. you dont need to upload the package lock json file
2. since there will be many files for exporting interfaces keeping all of 'em in types/scope.ts would make that file very large and wont be easy to read, so better keep that file as data.types.ts in the data folder itself

ThatDeparted2061 commented:
kindly work forward on the PRs as per the previous PRs that have been raised in for the TS integration
for eg: you can convert save.js next in the data folder

mahmoodalisha commented:
@ThatDeparted2061 Thanks for the suggestions for me to improve, I will keep Scope.ts file as data.types.ts in the data folder itself and that will be imported and reused in backupCircuit.ts only. Now I will work on save.js

Arnabdaz reviewed (changes requested):
- [x] please remove the `package-lock.json` from your pr.
- [x] put a proper title for the pr.
- [x] make the pr description more readable.

Arnabdaz committed:
Merge branch 'main' into mahmoodalisha/patch-79622

aryanndwi123 committed:
Merge branch 'main' into mahmoodalisha/patch-79622

aryanndwi123 committed:
Refactor backup function to eliminate redundant code

Removed disconnection of subcircuits before restoring connections. (as done in the .js file)

aryanndwi123 reviewed (approved):
LGTM 🍀

aryanndwi123 committed:
Merge branch 'main' into mahmoodalisha/patch-79622
